File: RefactoringTools/RefactoringTools/CodeRefactoringProvider.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CodeActions;
using Microsoft.CodeAnalysis.CodeRefactorings;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Rename;
using Microsoft.CodeAnalysis.Text;

namespace RefactoringTools
{
    [ExportCodeRefactoringProvider(CodeRefactoringProvider.RefactoringId, LanguageNames.CSharp)]
    internal class CodeRefactoringProvider : ICodeRefactoringProvider
    {
        public const string RefactoringId = "RefactoringTools";

        public async Task<IEnumerable<CodeAction>> GetRefactoringsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
        {
            // TODO: Replace the following code with your own analysis, generating a CodeAction for each refactoring to offer

            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);

            // Find the node at the selection.
            var node = root.FindNode(textSpan);

            // Only offer a refactoring if the selected node is a type declaration node.
            var typeDecl = node as TypeDeclarationSyntax;
            if (typeDecl == null)
            {
                return null;
            }

            // For any type declaration node, create a code action to reverse the identifier text.
            var action = CodeAction.Create("Reverse type name", c => ReverseTypeNameAsync(document, typeDecl, c));

            // Return this code action.
            return new[] { action };
        }

        private async Task<Solution> ReverseTypeNameAsync(Document document, TypeDeclarationSyntax typeDecl, CancellationToken cancellationToken)
        {
            // Produce a reversed version of the type declaration's identifier token.
            var identifierToken = typeDecl.Identifier;
            var newName = new string(identifierToken.Text.Reverse().ToArray());

            // Get the symbol representing the type to be renamed.
            var semanticModel = await document.GetSemanticModelAsync(cancellationToken);
            var typeSymbol = semanticModel.GetDeclaredSymbol(typeDecl, cancellationToken);

            // Produce a new solution that has all references to that type renamed, including the declaration.
            var originalSolution = document.Project.Solution;
            var optionSet = originalSolution.Workspace.Options;
            var newSolution = await Renamer.RenameSymbolAsync(document.Project.Solution, typeSymbol, newName, optionSet, cancellationToken).ConfigureAwait(false);

            // Return the new solution with the now-uppercase type name.
            return newSolution;
        }
    }
}

File: RefactoringTools/RefactoringTools.VB/CodeRefactoringProvider.vb
Imports Microsoft.CodeAnalysis.Rename

<ExportCodeRefactoringProvider(CodeRefactoringProvider.RefactoringId, LanguageNames.VisualBasic)>
Friend Class CodeRefactoringProvider
    Implements ICodeRefactoringProvider

    Public Const RefactoringId As String = "RefactoringTools.VB"

    Public Async Function GetRefactoringsAsync(document As Document, textSpan As TextSpan, cancellationToken As CancellationToken) As Task(Of IEnumerable(Of CodeAction)) _
        Implements ICodeRefactoringProvider.GetRefactoringsAsync

        ' TODO: Replace the following code with your own analysis, generating a CodeAction for each refactoring to offer

        Dim root = Await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(False)

        ' Find the node at the selection.
        Dim node = root.FindNode(textSpan)

        ' Only offer a refactoring if the selected node is a type statement node.
        Dim typeDecl = TryCast(node, TypeStatementSyntax)
        If typeDecl Is Nothing Then
            Return Nothing
        End If

        ' For any type statement node, create a code action to reverse the identifier text.
        Dim action = CodeAction.Create("Reverse type name", Function(c) ReverseTypeNameAsync(document, typeDecl, c))

        ' Return this code action.
        Return {action}
    End Function

    Private Async Function ReverseTypeNameAsync(document As Document, typeStmt As TypeStatementSyntax, cancellationToken As CancellationToken) As Task(Of Solution)
        ' Produce a reversed version of the type statement's identifier token.
        Dim identifierToken = typeStmt.Identifier
        Dim newName = New String(identifierToken.Text.Reverse().ToArray())

        ' Get the symbol representing the type to be renamed.
        Dim semanticModel = Await document.GetSemanticModelAsync(cancellationToken)
        Dim typeSymbol = semanticModel.GetDeclaredSymbol(typeStmt, cancellationToken)

        ' Produce a new solution that has all references to that type renamed, including the declaration.
        Dim originalSolution = document.Project.Solution
        Dim optionSet = originalSolution.Workspace.Options
        Dim newSolution = Await Renamer.RenameSymbolAsync(document.Project.Solution, typeSymbol, newName, optionSet, cancellationToken).ConfigureAwait(False)

        ' Return the new solution with the now-uppercase type name.
        Return newSolution
    End Function
End Class
